src/main.c

```diff
@@ -0,0 +1,257 @@
+#include <sys/epoll.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <arpa/inet.h>
+
+#include "lib.h"
+
+/*
+ * Write n bytes from buf to the provided fd, retrying short writes until
+ * we finish or hit an error. Assumes fd is blocking and therefore doesn't
+ * handle EAGAIN.
+ */
+void
+write_all(int fd, const char *buf, int n)
+{
+  int m = 0;
+  while(n > 0) {
+    m = write(fd, buf, n);
+    if(m < 0) {
+      perror("writing to stdout");
+      exit(1);
+    }
+    if(m == 0) {
+      fprintf(stderr, "early EOF when writing to stdout\n");
+      exit(1);
+    }
+    n -= m;
+  }
+}
+
+/*
+ * Connect to the given hostname on port 443 and return the file descriptor of
+ * the socket. On error, print the error and return -1. Caller is responsible
+ * for closing socket.
+ */
+int
+make_conn(const char *hostname)
+{
+  struct addrinfo *getaddrinfo_output, *rp;
+  int getaddrinfo_result =
+    getaddrinfo(hostname, "443", NULL, &getaddrinfo_output);
+  if(getaddrinfo_result != 0) {
+    fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(getaddrinfo_result));
+    return -1;
+  }
+
+  int sockfd = socket(getaddrinfo_output->ai_family,
+                      getaddrinfo_output->ai_socktype,
+                      getaddrinfo_output->ai_protocol);
+  if(sockfd < 0) {
+    perror("making socket");
+    return -1;
+  }
+
+  int connect_result = connect(
+    sockfd, getaddrinfo_output->ai_addr, getaddrinfo_output->ai_addrlen);
+  if(connect_result < 0) {
+    perror("connecting");
+    return -1;
+  }
+  freeaddrinfo(getaddrinfo_output);
+  return sockfd;
+}
+
+/*
+ * Given an established TCP connection, and a rustls client_session, send an
+ * HTTP request and read the response. On success, return 0. On error, print
+ * the message and return 1.
+ */
+int
+send_request_and_read_response(int sockfd, void *client_session,
+                               const char *hostname, const char *path)
+{
+  char buf[2048];
+
+  memset(buf, '0', sizeof(buf));
+  snprintf(buf,
```

Comment:
nit: you don't get `snprintf(3)` in C89, but that's irrelevant for this test program which isn't expected to compile in all the places `curl` does.